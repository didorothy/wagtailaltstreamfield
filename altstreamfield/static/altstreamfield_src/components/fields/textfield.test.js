import React from "react";
import { render, fireEvent, wait, waitForDomChange, waitForElementToBeRemoved, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import TextField from "./textfield";
import FormErrorContext from "../../context/formerror";

describe('TextField', () => {
    let autosize;
    let jquery;
    beforeEach(() => {
        autosize = window.autosize;
        jquery = window.jQuery;

        window.autosize = function() {};
        window.jQuery = function() {};
    });

    afterEach(() => {
        window.autosize = autosize;
        window.jQuery = jquery;
    });

    test('#minimal', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value="sample text" name="test" onChange={() => {}} />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('#label', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value="sample text" name="test" onChange={() => {}} label="Test Text Field" />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('#help_text', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value="sample text" name="test" onChange={() => {}} label="Test Text Field" help_text="Sample help text." />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('#required', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value="sample text" name="test" onChange={() => {}} label="Test Text Field" help_text="Sample help text." required={true} />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('#error', () => {
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value="" name="test" onChange={() => {}} label="Test Text Field" help_text="Sample help text." required={true} />
        </FormErrorContext.Provider>);
        expect(container.container).toMatchSnapshot();
        cleanup();
    });

    test('min_length', () => {
        let current_value = "";
        let onChange = (name, value) => {
            current_value = value;
        }
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Text Field" help_text="Sample help text." required={true} min_length={3}/>
        </FormErrorContext.Provider>);
        expect(manager.isValid()).toEqual(false);
        let input = container.container.getElementsByTagName('textarea')[0];
        let error = container.container.getElementsByClassName('error-message')[0];
        expect(error.textContent).toEqual('This field is required.')

        fireEvent.change(input, {target: {value: 'a'}});
        container.rerender(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Text Field" help_text="Sample help text." required={true} min_length={3}/>
        </FormErrorContext.Provider>);
        expect(current_value).toEqual('a');
        expect(manager.isValid()).toEqual(false);

        fireEvent.change(input, {target: {value: 'abc'}});
        container.rerender(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Text Field" help_text="Sample help text." required={true} min_length={3}/>
        </FormErrorContext.Provider>);
        expect(current_value).toEqual('abc');
        expect(manager.isValid()).toEqual(true);
        cleanup();
    });

    test('max_length', () => {
        let current_value = "";
        let onChange = (name, value) => {
            current_value = value;
        }
        let manager = new FormErrorContext.FormErrorManager();
        let container = render(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Text Field" help_text="Sample help text." required={true} max_length={3}/>
        </FormErrorContext.Provider>);
        expect(manager.isValid()).toEqual(false);
        let input = container.container.getElementsByTagName('textarea')[0];
        let error = container.container.getElementsByClassName('error-message')[0];
        expect(error.textContent).toEqual('This field is required.')

        fireEvent.change(input, {target: {value: 'a'}});
        container.rerender(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Text Field" help_text="Sample help text." required={true} max_length={3}/>
        </FormErrorContext.Provider>);
        expect(current_value).toEqual('a');
        expect(manager.isValid()).toEqual(true);

        fireEvent.change(input, {target: {value: 'abcd'}});
        container.rerender(<FormErrorContext.Provider value={manager}>
            <TextField owner_id="struct-01" value={current_value} name="test" onChange={onChange} label="Test Text Field" help_text="Sample help text." required={true} max_length={3}/>
        </FormErrorContext.Provider>);
        expect(current_value).toEqual('abcd');
        expect(manager.isValid()).toEqual(false);
        cleanup();
    });
});