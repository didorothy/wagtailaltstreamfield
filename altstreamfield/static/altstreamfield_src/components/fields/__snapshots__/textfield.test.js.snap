// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField #error 1`] = `
<div>
  <div
    class="field text_field required error"
  >
    <label
      for="struct-01test"
    >
      Test Text Field
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <textarea
          cols="40"
          id="struct-01test"
          placeholder="Test Text Field"
          rows="1"
        />
      </div>
      <p
        class="help"
      >
        Sample help text.
      </p>
      <p
        class="error-message"
      >
        This field is required.
      </p>
    </div>
  </div>
</div>
`;

exports[`TextField #help_text 1`] = `
<div>
  <div
    class="field text_field"
  >
    <label
      for="struct-01test"
    >
      Test Text Field
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <textarea
          cols="40"
          id="struct-01test"
          placeholder="Test Text Field"
          rows="1"
        >
          sample text
        </textarea>
      </div>
      <p
        class="help"
      >
        Sample help text.
      </p>
      
    </div>
  </div>
</div>
`;

exports[`TextField #label 1`] = `
<div>
  <div
    class="field text_field"
  >
    <label
      for="struct-01test"
    >
      Test Text Field
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <textarea
          cols="40"
          id="struct-01test"
          placeholder="Test Text Field"
          rows="1"
        >
          sample text
        </textarea>
      </div>
      
    </div>
  </div>
</div>
`;

exports[`TextField #minimal 1`] = `
<div>
  <div
    class="field text_field"
  >
    <label
      for="struct-01test"
    >
      Test
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <textarea
          cols="40"
          id="struct-01test"
          placeholder="Test"
          rows="1"
        >
          sample text
        </textarea>
      </div>
      
    </div>
  </div>
</div>
`;

exports[`TextField #required 1`] = `
<div>
  <div
    class="field text_field required"
  >
    <label
      for="struct-01test"
    >
      Test Text Field
    </label>
    <div
      class="field-content"
    >
      <div
        class="input"
      >
        <textarea
          cols="40"
          id="struct-01test"
          placeholder="Test Text Field"
          rows="1"
        >
          sample text
        </textarea>
      </div>
      <p
        class="help"
      >
        Sample help text.
      </p>
      
    </div>
  </div>
</div>
`;
